### PR TITLE
Bootstrap Nomad ACL system if no token is given

### PR DESCRIPTION
**Pull request by Mongey**
Similar to the [Bootstrap the Consul ACL system if no token is given][boostrap-consul]
it would be very useful to bootstrap Nomads ACL system and manage it in
Vault.

[boostrap-consul]:https://github.com/hashicorp/vault/pull/10751

### CODE REVIEW

**Comment by hghaf099** [comment]
@Mongey  are you still interested in working on this PR?

**Comment by Mongey** [comment]
@hghaf099 yep

**Comment by Mongey** [comment]
bump

File: builtin/logical/nomad/backend_test.go [hunk 1]
```diff
@@ -57,24 +64,61 @@ func prepareTestContainer(t *testing.T) (func(), *Config) {
...
+			ServiceURL: *u,
+			Token:      nomadToken,
+		}, nil
+	})
+	if err != nil {
+		t.Fatalf("Could not start docker Nomad: %s", err)
+	}
+
+	// the container may not be ready yet ... wait a bit
+	time.Sleep(time.Second * 15)
```
Comment (swenson) [inline builtin/logical/nomad/backend_test.go:104]: Why did this become necessary?

File: builtin/logical/nomad/backend_test.go [hunk 2]
```diff
@@ -27,7 +28,13 @@ func (c *Config) APIConfig() *nomadapi.Config {
...
 }
 
-func prepareTestContainer(t *testing.T) (func(), *Config) {
+func (c *Config) Client() (*nomadapi.Client, error) {
+	apiConfig := c.APIConfig()
+
+	return nomadapi.NewClient(apiConfig)
+}
+
+func prepareTestContainer(t *testing.T, manuallyBootstrap bool) (func(), *Config) {
```
Comment (swenson) [inline builtin/logical/nomad/backend_test.go:37]: nit: It is not immediately clear to me what "manually" means in this context. I assume it means the opposite of "automatic".

Would it be better to invert it and make it `autoACL` or something similar, perhaps?

File: builtin/logical/nomad/path_config_access.go [hunk 3]
```diff
@@ -129,6 +129,20 @@ func (b *backend) pathConfigAccessWrite(ctx context.Context, req *logical.Reques
 		conf.ClientKey = clientKey.(string)
 	}
 
+	fmt.Printf("[WARN] Token is %v", conf.Token)
```
Comment (swenson) [inline builtin/logical/nomad/path_config_access.go:132]: We'll definitely want to remove this debugging line :)

File: builtin/logical/nomad/backend_test.go [hunk 4]
```diff
@@ -57,24 +64,61 @@ func prepareTestContainer(t *testing.T) (func(), *Config) {
...
+			nomadAuth, err := nomadapi.NewClient(nomadAuthConfig)
+			if err != nil {
+				return nil, err
+			}
+
+			err = preprePolicies(nomadAuth)
+			if err != nil {
+				return nil, err
+			}
+		} else {
```
Comment (swenson) [inline builtin/logical/nomad/backend_test.go:90]: nit: no need for this `else`? Or was something intended to go here?

**Review by swenson — COMMENTED** [review]
Thanks for this! Just a few small-ish comments.

File: builtin/logical/nomad/backend_test.go (hunk 2 above)
Comment (swenson) [reply builtin/logical/nomad/backend_test.go:37]: In light of https://github.com/hashicorp/vault/pull/10751 -- maybe naming it `bootstrap` would be preferable.

File: builtin/logical/nomad/backend_test.go (hunk 1 above)
Comment (Mongey) [reply builtin/logical/nomad/backend_test.go]: The container wasn't ready to accept connections, but I now see that `NewServiceRunner` will retry until we get a non-error, so I added a call to get the nomad leader in the setup

**Review by swenson — APPROVED** [review]
LGTM

I pulled it down and tested it locally. It seems to simplify the process of getting it to work locally, i.e., I can now just run

```sh
# window 1
sudo nomad agent -dev -bind 0.0.0.0 -log-level INFO -acl-enabled
# window 2
vault server -dev 
# window 3
vault secrets enable nomad
vault write nomad/config/access address=http://127.0.0.1:4646
vault write nomad/role/monitoring policies=readonly
vault read nomad/creds/monitoring
```

And skip having to run `nomad acl boostrap`. Though if we do still run `nomad acl boostrap` manually, as before, it works as expected.

Thanks so much for your contribution!

**Pull request merged**